bin/dynamic_challenge_send_email.py

```diff
@@ -0,0 +1,206 @@
+            "INVALID",
+            "no",
+        ): f"Evaluation failed for Submission {submission_id}."
+        + "\n"
+        + f"Reason: '{reason}'."
+        + "\n"
+        + "Please contact the organizers for more information.",
+    }
+
+    body = templates.get((status, email_with_score))
+
+    # If there is a typo in ``email_with_score``, ``body`` will be None;
+    # Raise an error if so, to avoid sending empty e-mails...
+    if body is None:
+        raise ValueError(
+            f"Incorrect status and/or email_with_score arguments. Got status: {status}, email_with_score: {email_with_score}."
+        )
+
+    return body
+
+
+def get_evaluation(syn: synapseclient.Synapse, submission_id: str) -> Tuple[str, str]:
+    """Get evaluation id for the submission
+
+    Arguments:
+        syn: Synapse connection
+        submission_id: The id of submission
+
+    Returns:
+        eval: the tuple of evaluation ID and evaluation name.
+
+    Raises:
+        Exception: if an error occurs
+    """
+    try:
+        eval_id = syn.getSubmission(submission_id, downloadFile=False).get(
+            "evaluationId"
+        )
+        eval_name = syn.getEvaluation(eval_id).get("name")
+        return eval_id, eval_name
+    except Exception as e:
+        print(
+            f"An error occurred while retrieving the evaluation for submission {submission_id}: {e}"
+        )
+
+
+def get_target_link(synapse_client: synapseclient.Synapse, eval_id: str) -> str:
+    """
+    Retrieves the redirection link returned in the email to view submissions for a given submission evaluation ID.
+
+    Arguments:
+        syn: Synapse connection
+        eval_id: the evaluation id of submission.
+
+    Returns:
+        link: The redirection link to the submission page.
+    """
+
+    EVAL_TO_LINK = {
+        "9615379": "https://www.synapse.org/#!Synapse:syn52052735/wiki/626195",
+        "9615532": "https://www.synapse.org/#!Synapse:syn52052735/wiki/626203",
+        "9615534": "https://www.synapse.org/#!Synapse:syn52052735/wiki/626211",
+        "9615535": "https://www.synapse.org/#!Synapse:syn52052735/wiki/626216",
+    }
+    link = EVAL_TO_LINK.get(eval_id, None)
+    if link:
+        return link
+    project_id = synapse_client.getEvaluation(eval_id).get("contentSource")
+    return f"https://www.synapse.org/#!Synapse:{project_id}"
+
+
+def send_email(
+    submission_id: str, email_with_score: str, notification_type: str
+) -> None:
+    """
+    Sends an e-mail on the status of the individual submission
+    to the submitting team or individual.
+
+    Arguments:
+      submission_id: The ID for an individual submission within an evaluation queue
+      email_with_score: Whether to include the score in the e-mail
+      notification_type: The type of notification to send
+
+    Raises:
+      ValueError: if an incorrect type is provided for notification_type
+    """
+    # Initiate connection to Synapse
+    syn = synapseclient.login()
```

Comment:
nit: notification_type is not one of the valid values this `.login()` doesn't need to be called. Consider adding an initial guard statement to the beginning of the function so you know you'll continue with one of the branches.

ie:
```
def send_email(
    submission_id: str, email_with_score: str, notification_type: str
) -> None:
    """
    Sends an e-mail on the status of the individual submission
    to the submitting team or individual.
    Arguments:
      submission_id: The ID for an individual submission within an evaluation queue
      email_with_score: Whether to include the score in the e-mail
      notification_type: The type of notification to send
    Raises:
      ValueError: if an incorrect type is provided for notification_type
    """
    if notification_type != AFTER and notification_type != BEFORE:
        raise ValueError(f"Invalid notification_type. Must be '{BEFORE}' or '{AFTER}'")
````